### PR TITLE
Implemented {Array, TypedArray}.prototype.{with, toReversed, toSorted} (#26, #27, #28, #29)

### PR DESCRIPTION
**Pull request by shiroyk**
`{Array, TypedArray}.prototype.toSorted` #26
`{Array, TypedArray}.prototype.toReversed` #27
`{Array, TypedArray}.prototype.with` #28
`Array.prototype.toSpliced` #29

### CODE REVIEW

**Comment by mstoykov** [comment]
Thank you for the PR @shiroyk :bow: I will try to review it within the week

File: builtin_array.go [hunk 1]
```diff
@@ -1224,6 +1224,169 @@ func (r *Runtime) arrayproto_flatMap(call FunctionCall) Value {
...
+		actualIndex = relativeIndex
+	} else {
+		actualIndex = length + relativeIndex
+	}
+	if actualIndex >= length || actualIndex < 0 {
+		panic(r.newError(r.getRangeError(), "Invalid index %s", call.Argument(0).String()))
+	}
+	a := r.newArrayLength(length)
+
+	for k := int64(0); k < length; k++ {
```
Comment (mstoykov) [inline builtin_array.go:1244]: I guess that if you are using the optimization below I would prefer that it done here as well.

File: builtin_array.go [hunk 2]
```diff
@@ -1224,6 +1224,169 @@ func (r *Runtime) arrayproto_flatMap(call FunctionCall) Value {
...
+func (r *Runtime) arrayproto_toSpliced(call FunctionCall) Value {
+	o := call.This.ToObject(r)
+	length := toLength(o.self.getStr("length", nil))
+	actualStart := relToIdx(call.Argument(0).ToInteger(), length)
+	var actualSkipCount int64
+	switch len(call.Arguments) {
+	case 0:
+	case 1:
+		actualSkipCount = length - actualStart
+	default:
```
Comment (mstoykov) [inline builtin_array.go:1323]: This being with an `if len(...) >= 2)` seems way more readable.

File: builtin_array.go [hunk 3]
```diff
@@ -1224,6 +1224,169 @@ func (r *Runtime) arrayproto_flatMap(call FunctionCall) Value {
...
+				values = src.values[:newLength]
+				copy(values[actualStart+itemCount:], values[actualStart+actualSkipCount:length])
+			} else {
+				values = make([]Value, newLength)
+				copy(values, src.values[:actualStart])
+				copy(values[actualStart+itemCount:], src.values[actualStart+actualSkipCount:])
+			}
+		} else {
+			values = src.values
+		}
```
Comment (mstoykov) [inline builtin_array.go:1353]: Given that the whole point of toSpliced is to not touch the original array. The code here seems very risky and IMO buggy, just not in a way that the tc39 test catch it.

Try to add a test such as 
```golang

func TestArrayToSpliced(t *testing.T) {
	const SCRIPT = `
		const a = [1, 2, 3];
		a.push(4)
		assert(compareArray(a, [1, 2, 3, 4]));
		const b = a.toSpliced(2)
		$ERROR(a)
		assert(compareArray(a, [2, 3, 4]));
		// assert(compareArray(b, [2, 3, 4, 3]));
	`
	testScriptWithTestLib(SCRIPT, _undefined, t)
}
```

And you will notice that `a` is now `1,2,,,` even though it is supposed to not be changed

**Review by mstoykov — CHANGES_REQUESTED** [review]

File: builtin_array.go (hunk 1 above)
Comment (shiroyk) [reply builtin_array.go:1244]: I will optimize this method.

File: builtin_array.go (hunk 3 above)
Comment (shiroyk) [reply builtin_array.go:1353]: Sorry, I forgot to check this part.

File: builtin_arrray_test.go [hunk 4]
```diff
@@ -339,3 +341,23 @@ func TestArrayProto(t *testing.T) {
...
+	assert(compareArray(a, [1, 2, 3, 4]));
+	assert(compareArray(b, [1, 2]));
+	a.push(5)
+	const c = a.toSpliced(1, 2);
+	assert(compareArray(a, [1, 2, 3, 4, 5]));
+	assert(compareArray(c, [1, 4, 5]));
+	assert(compareArray(a.toSpliced(4, 2, 'a', 'b'), [1, 2, 3, 4, 'a', 'b']));
+	assert(compareArray(a.toSpliced(-2, 2), [1, 2, 3]));
+	assert(compareArray(a.toSpliced(2, 10), [1, 2]));
+	assert(compareArray(a.toSpliced(1, 0, 'a'), [1, 'a', 2, 3, 4, 5]));
```
Comment (mstoykov) [inline builtin_arrray_test.go:360]: Can you please check that `a` doesn't change in between this as well

File: builtin_array.go [hunk 5]
```diff
@@ -1224,6 +1224,188 @@ func (r *Runtime) arrayproto_flatMap(call FunctionCall) Value {
...
+		panic(r.NewTypeError("Invalid array length"))
+	}
+
+	if src := r.checkStdArrayObj(o); src != nil {
+		var values []Value
+		if itemCount < actualSkipCount {
+			values = make([]Value, len(src.values))
+			copy(values, src.values)
+			copy(values[actualStart+itemCount:], values[actualStart+actualSkipCount:])
+			values = values[:newLength]
```
Comment (mstoykov) [inline builtin_array.go:1358]:
```suggestion
			values = make([]Value, newLength)
			copy(values, src.values[actualStart])
			copy(values[actualStart+itemCount:], src.values[actualStart+actualSkipCount:])
```

Haven't checked the code - but there is no need to overt allocate to then cut down.

File: builtin_array.go [hunk 6]
```diff
@@ -1224,6 +1224,188 @@ func (r *Runtime) arrayproto_flatMap(call FunctionCall) Value {
...
+		} else if itemCount > actualSkipCount {
+			if int64(cap(src.values)) >= newLength {
+				values = make([]Value, newLength)
+				copy(values, src.values[:actualStart])
+				copy(values[actualStart+itemCount:], src.values[actualStart+actualSkipCount:])
+			} else {
+				values = make([]Value, newLength)
+				copy(values, src.values[:actualStart])
+				copy(values[actualStart+itemCount:], src.values[actualStart+actualSkipCount:])
+			}
```
Comment (mstoykov) [inline builtin_array.go:1368]: This code seems to be the same

Comment (mstoykov) [reply builtin_array.go:1368]: and arguablt is the same as the code I propose above, so it seems like we do not need all of those ifs.

**Review by mstoykov — CHANGES_REQUESTED** [review]

**Review by mstoykov — APPROVED** [review]
LGTM! 

Sorry for the slow reviews, but this kind of aligned with vacation time and us releasing a new k6 version.

We are currently trying to keep Sobek and [goja](https://github.com/dop251/goja) as close as possible.

Do you want to make a PR to add this to goja as well or do you want me to do it for you?

**Comment by mstoykov** [comment]
Ah ... unfortunately me merging some updates from goja made conflicts wiht tc39_test.go - can you please fix them

**Comment by shiroyk** [comment]
OK, I will add this PR to Goja.

**Comment by mstoykov** [comment]
Thank you for all the work @shiroyk :bow: 

I merged the upstream changes  in #42  and will close this PR

**Pull request closed without merging**